File: src/store/factory.py
"""Store construction — the one place that names a concrete adapter.

Adding a database is: write an adapter implementing `store.base.Store`, then add
one line here. Nothing above the store seam changes.
"""
from __future__ import annotations

from ..errors import ConfigError
from .base import Store


def make_store(kind: str, host: str, port: int, database: str,
               options: dict | None = None) -> Store:
    """Build the configured store. `options` carries adapter-specific settings
    (e.g. ClickHouse's `async_insert`); unknown keys are the adapter's business."""
    opts = options or {}
    if kind == "clickhouse":
        from .clickhouse import ClickHouseStore
        return ClickHouseStore(host, port, database,
                               bool(opts.get("async_insert", False)))
    if kind == "memory":
        from .memory import MemoryStore
        return MemoryStore()
    raise ConfigError(
        f"config: unknown store.kind '{kind}' (known: clickhouse, memory)")

File: internal/store/factory/factory.go
// Package factory is the one place a concrete store adapter is named.
//
// Adding a database is: write an adapter implementing store.Store, then add one
// case here. Nothing above the store seam changes.
package factory

import (
	"github.com/vishal9359/Analysis-FW/internal/fwerr"
	"github.com/vishal9359/Analysis-FW/internal/store"
	"github.com/vishal9359/Analysis-FW/internal/store/clickhouse"
	"github.com/vishal9359/Analysis-FW/internal/store/memory"
)

// New builds the configured store. opts carries adapter-specific settings (e.g.
// ClickHouse's async_insert); unknown keys are the adapter's business.
func New(kind, host string, port int, database string, opts map[string]interface{}) (store.Store, error) {
	switch kind {
	case "clickhouse":
		async := false
		if v, ok := opts["async_insert"].(bool); ok {
			async = v
		}
		return clickhouse.New(host, port, database, async), nil
	case "memory":
		return memory.New(), nil
	default:
		return nil, fwerr.Config("config: unknown store.kind %q (known: clickhouse, memory)", kind)
	}
}
